src/utils/load-audio.tsx

```diff
@@ -5,10 +5,32 @@ import { renderAsEntity } from "../utils/jsx-entity";
 import { loadAudioTexture } from "../utils/load-audio-texture";
 import { HubsWorld } from "../app";
 import { HubsVideoTexture } from "../textures/HubsVideoTexture";
+import { EntityID } from "./networking-types";
+import { MediaVideo } from "../bit-components";
+import { VIDEO_FLAGS } from "../inflators/video";
+
+export function* loadAudio(world: HubsWorld, eid: EntityID, url: string) {
+  let loop = true;
+  let autoPlay = true;
+  let controls = true;
+  if (MediaVideo.flags[eid]) {
+    loop = (MediaVideo.flags[eid] & VIDEO_FLAGS.LOOP) !== 0;
+    autoPlay = (MediaVideo.flags[eid] & VIDEO_FLAGS.AUTOPLAY) !== 0;
+    controls = (MediaVideo.flags[eid] & VIDEO_FLAGS.CONTROLS) !== 0;
+  } else {
+    MediaVideo.flags[eid] |= VIDEO_FLAGS.AUTOPLAY;
+    MediaVideo.flags[eid] |= VIDEO_FLAGS.LOOP;
+    MediaVideo.flags[eid] |= VIDEO_FLAGS.CONTROLS;
+  }
+  let projection = ProjectionMode.FLAT;
+  if (MediaVideo.projection[eid]) {
+    projection = APP.getString(MediaVideo.projection[eid]) as ProjectionMode;
+  } else {
+    MediaVideo.projection[eid] = APP.getSid(ProjectionMode.FLAT);
+  }
```

Comment:
I don't get these `MediaVideo.flags[eid]` and `MediaVideo.projection[eid]` checking are for. Does it mean an entity has `MediaVideo` component but thoese properties are not set? If so, how does it happen?

Reply:
Yeah this is very confusing. There are two path how we load images (same for the rest of media):
- Spawned Media:
**MediaPrefab -> inflateMediaLoader -> [media-loading-system] -> loadImage -> inflateImage (MediaImage component added)**
- glTF:
**inflateImageLoader -> inflateMediaLoader -> [media-loading-system] -> loadImage -> inflateImage (MediaImage component added)**

I am using `MediaImage` as a transport container for the `ImageLoaderParams` that are later gathered in `inflateImageLoader` and used in `loadImage` as it assumes that the `MediaImage` has been added an initialized already at that point. As you pointed out this is not correct as the `MediaImage` component has not been added yet.

I'm going to introduce a `MediaImageLoaderData` and `MediaVideoLoaderData` data stores in between `inflateImageLoader` and `loadImage` to safely pass the initialization parameters for all medias. I think that would solve most of the concerns in this PR.

Reply:
Thanks for explaining it. Give me a little time to think whether we can simplify further...

Reply:
Probably the root cause of the complexity is MediaLoader system is too big and operate too many things. If media type specific operation is needed, they can't be decoupled well from MediaLoader system.

For example, We may separate media loder system into media type specific loader system like ImageLoaderSystem, VideoLoaderSystem, ModelLoaderSystem, and common operations can be exported as utils that can be used from the new loaders. Each loader would be simplerer.

But I don't think we have time to refactor. So transporting as temporary solution would be ok.

The current `MediaImageLoaderData` and `MediaVideoLoaderData` approach may be a bit less maintainable. They are as if just global variables and future readers may feel hard to understand how they are properly managed. And it is decoupled from bitecs component and entity mechanism. So for example even if an component or entity is removed while the operation, the data is remained. It can be error prone, especially if the entity is recycled.

Perhaps managing the transporting data as component is safer? Give me more time for thinking proper temporary solution...

Reply:
Clean up idea https://github.com/mozilla/hubs/commit/3d2f38102ac052c1a7fa7834a02745d3b47bf750

* Treat MediaImage/VideoLoaderData as part of MediaLoader component data. Less problematic because their life cycle will be synched with MediaLoader component.
* Remove dependency with entity and MediaImage/VideoLoaderData from loaders in src/utils. The loaders will be simpler.

Feel free to cherry-pick if it looks good, or I can make a follow up PR after this PR lands.

Reply:
Thanks, sounds good to me. Maybe better to cleanup this after landing. Feel free to update that when this lands and I'll review.